script/src/verify.rs

```diff
@@ -1029,50 +1026,159 @@ impl<DL: CellDataProvider + HeaderProvider + ExtensionProvider + Send + Sync + C
+// Run a series of VMs that are just freshly resumed
+fn run_vms(
+    script_group: &ScriptGroup,
+    max_cycles: Cycle,
+    mut machines: Vec<(Machine, ResumeData)>,
+    context: &Arc<Mutex<MachineContext>>,
+) -> Result<ChunkState, ScriptError> {
+    // (exit_code, cycles, resume_data)
+    let mut callee_data = None;
+
+    let map_vm_internal_error = |error: VMInternalError| match error {
+        VMInternalError::CyclesExceeded => ScriptError::ExceededMaximumCycles(max_cycles),
+        _ => ScriptError::VMInternalError(format!("{error:?}")),
+    };
+
+    while !machines.is_empty() {
+        let (mut machine, resume_data) = machines.pop().unwrap();
+
+        if let Some((callee_exit_code, callee_cycles, callee_resume_data)) = callee_data {
+            update_caller_machine(
+                &mut machine.machine,
+                callee_exit_code,
+                callee_cycles,
+                &callee_resume_data,
+            )
+            .map_err(map_vm_internal_error)?;
         }
 
         match machine.run() {
             Ok(code) => {
-                if code == 0 {
-                    Ok(ChunkState::Completed(machine.machine.cycles()))
-                } else {
-                    Err(ScriptError::validation_failure(&script_group.script, code))
-                }
+                callee_data = Some((code, machine.machine.cycles(), resume_data));
             }
             Err(error) => match error {
                 VMInternalError::CyclesExceeded => {
-                    Ok(ChunkState::suspended(ResumableMachine::new(machine, None)))
+                    let mut new_suspended_machines: Vec<_> = {
+                        let mut context = context.lock().map_err(|e| {
+                            ScriptError::VMInternalError(format!("Failed to acquire lock: {}", e))
+                        })?;
+                        context.suspended_machines.drain(..).collect()
+                    };
+                    // The inner most machine lives at the top of the vector,
+                    // reverse the list for natural order.
+                    new_suspended_machines.reverse();
+                    let mut suspended_machines: Vec<_> = machines
+                        .drain(..)
+                        .map(|(machine, data)| ResumableMachine::new(machine, data))
+                        .collect();
+                    suspended_machines.push(ResumableMachine::new(machine, resume_data));
+                    suspended_machines.append(&mut new_suspended_machines);
+                    return Ok(ChunkState::suspended(
+                        suspended_machines,
+                        Arc::clone(context),
+                    ));
                 }
-                _ => Err(ScriptError::VMInternalError(format!("{error:?}"))),
+                _ => return Err(ScriptError::VMInternalError(format!("{error:?}"))),
             },
-        }
+        };
+    }
+
+    let (exit_code, cycles, _) = callee_data.unwrap();
```

Comment:
Is it possible for `callee_data` to be `None`?

Reply:
Ideally when we enter this function, we are expecting at least one VM instance, which will ensure `callee_data` is not `None`. I've added a check at the top of this function ensuring at least one VM exists.